data_structures/src/mainnet_validations.rs

```diff
@@ -16,6 +18,143 @@ pub const FIRST_HARD_FORK: Epoch = 192000;
 /// 28 April 2021 @ 9:00:00 UTC
 pub const SECOND_HARD_FORK: Epoch = 376320;
 
+/// TAPI Engine
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct TapiEngine {
+    /// bit votes counter by bits
+    pub bit_tapi_counter: BitTapiCounter,
+    /// wip activation
+    pub wip_activation: HashMap<String, Epoch>,
+}
+
+impl TapiEngine {
+    pub fn update_bit_counter(&mut self, v: u32, epoch: Epoch, avoid_wip_list: &HashSet<String>) {
+        // In case of empty epochs, they would be considered as blocks with tapi version to 0
+        if self.bit_tapi_counter.last_epoch != 0 && epoch > self.bit_tapi_counter.last_epoch + 1 {
```

Comment:
What happens if `bit_tapi_counter.last_epoch == 0`?

Reply:
that it would be updated as a normal update, and last_epoch would be overwrite with the last consolidated block epoch